pkg/azure/aks/pricing_map.go

```diff
@@ -0,0 +1,183 @@
+package aks
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/Azure/go-autorest/autorest/to"
+	retailPriceSdk "gomodules.xyz/azure-retail-prices-sdk-for-go/sdk"
+)
+
+const (
+	AZ_API_VERSION string = "2023-01-01-preview" // using latest API Version https://learn.microsoft.com/en-us/rest/api/cost-management/retail-prices/azure-retail-prices
+)
+
+type MachineOperatingSystem int
+
+const (
+	Linux MachineOperatingSystem = iota
+	Windows
+)
+
+func (o MachineOperatingSystem) String() string {
+	return [2]string{"Linux", "Windows"}[o-1]
+}
+
+type MachinePriority int
+
+const (
+	OnDemand MachinePriority = iota
+	Spot
+)
+
+func (v MachinePriority) String() string {
+	return [2]string{"OnDemand", "Spot"}[v-1]
+}
+
+type PriceBySku map[string]retailPriceSdk.ResourceSKU
+
+type PriceByOperatingSystem map[MachineOperatingSystem]PriceBySku
+
+type PriceByPriority map[MachinePriority]PriceByOperatingSystem
+
+type PriceStore struct {
+	lock              *sync.RWMutex
+	subscriptionId    string
+	logger            *slog.Logger
+	context           context.Context
+	retailPriceClient *retailPriceSdk.RetailPricesClient
+
+	RegionMap map[string]PriceByPriority
+	Cache     map[string]*retailPriceSdk.ResourceSKU
+}
+
+func NewPricingStore(subId string, priceClient *retailPriceSdk.RetailPricesClient, parentLogger *slog.Logger, parentContext context.Context) *PriceStore {
+	logger := parentLogger.With("subsystem", "pricingMap")
+
+	p := &PriceStore{
+		lock:              &sync.RWMutex{},
+		logger:            logger,
+		context:           parentContext,
+		subscriptionId:    subId,
+		retailPriceClient: priceClient,
+
+		RegionMap: make(map[string]PriceByPriority),
+		Cache:     make(map[string]*retailPriceSdk.ResourceSKU),
+	}
+
+	go func() {
+		err := p.PopulatePriceStore([]string{})
+		if err != nil {
+			p.logger.LogAttrs(p.context, slog.LevelError, "error populating initial price store", slog.String("error", err.Error()))
+		}
+	}()
```

Comment:
I'm trying to understand what specifically this is doing. You're passing in an empty slice to the PopulatePriceStore, what's the significance of that? If there's an error, is the assumption the next tick will populate the pricing store? IMO I think it would be better to
1. retry N times
2. If you can't initialize the pricing store, it should return an error and halt the program

This would prevent deployments from continuing if the pricing store can never be initialized.

Reply:
Good question!

> You're passing in an empty slice to the PopulatePriceStore, what's the significance of that?

The location list is not used upon initial population because we don't know which locations are relevant, so we just get _everything_.  This is super inefficient as it is very unlikely any users will have a presence in every Azure region, and almost all space and compute time building the map will never be accessed.

When we have computed a list of VMs (implemented in a later PR), we will know the regions that the user actually has a presence in, so we can drastically reduce the API call to just the relevant regions for the user.  In the initial run of this, we need everything, but any subsequent run should have an idea of the regions that are necessary.

```
 If there's an error, is the assumption the next tick will populate the pricing store? IMO I think it would be better to

- retry N times
- If you can't initialize the pricing store, it should return an error and halt the program
This would prevent deployments from continuing if the pricing store can never be initialized.
```

I think this would be a good future improvement. The way that I'm envisioning this is that there is a background job that populates the map every so often, and will crash the exporter if it can't populate.

I've added those as TODO items in the README.